### ISSUE TITLE
swarm stack deploy volume name not allowed

### ISSUE THREAD

**Issue by Joyam** [moby/moby#35111]
**Description**

Unable to deploy stack with a volume name

**Steps to reproduce the issue:**
1. 
create docker-compose.yml like that :
```yml
version: "3"

services:
  apache:
    image: "apache"
    ports:
      - "80:80"
      - "443:443"
    volumes:
      - vol-name:/var/www/project

volumes:
  vol-name:
    external:
      name: other-volume-name
```

2. `docker-compose config > foo.yml`
3. `docker stack deploy -c foo.yml bar`

**Describe the results you received:**
name Additional property name is not allowed

**Describe the results you expected:**
The stack deploys successfully

**Additional information you deem important (e.g. issue happens only occasionally):**
generated `foo.yml` file looks like this :
```yml
services:
  apache:
    image: apache
    ports:
    - 80:80/tcp
    - 443:443/tcp
    volumes:
    - vol-name:/var/www/project:rw
version: '3.0'
volumes:
  vol-name:
    external: true
    name: other-volume-name
```
In this case we have the following error : `name Additional property name is not allowed`

and if I make it look like this : 

```yml
services:
  apache:
    image: apache
    ports:
    - 80:80/tcp
    - 443:443/tcp
    volumes:
    - vol-name:/var/www/project:rw
version: '3.0'
volumes:
  vol-name:
    external:
      name: other-volume-name
```

it works but generates this warning :
`WARN[0000] volume vol-name: volume.external.name is deprecated in favor of volume.name`

**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:09 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:48 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 7
 Running: 3
 Paused: 0
 Stopped: 4
Images: 31
Server Version: 17.09.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: wr2qxb1m0bjbubhbd9uf87vej
 Is Manager: true
 ClusterID: w74skx7rusf129ubuh4lolmeh
 Managers: 3
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 10.1.124.1
 Manager Addresses:
  10.1.124.1:2377
  10.1.124.2:2377
  10.1.124.3:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.0-3-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 3.863GiB
Name: mynode
ID: O4L2:AXOS:H5AI:35TC:ASWA:2JX2:B556:UQSR:AIS7:26LI:624R:45XH
Docker Root Dir: /home/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

```
$ docker-compose --version
docker-compose version 1.16.1, build 6d1ac21
```

**Comment by dnephin** [comment]
You'll have to upgrade to version 3.4 to use `volume.name`, which is the new recommended field.

I've created https://github.com/docker/cli/issues/608 to fix the warning in versions before 3.4.  If you'd like to stick with a version before 3.4 for some reason then continue to use volume.external.name and disregard the warning.

**Comment by dnephin** [comment]
Closing the issue here since the problem is in `docker/cli`

**Comment by dinvlad** [comment]
Not sure if this is related, but when we use a 3.4 compose, the Swarm ignores `volume.name` altogether (i.e. the volume is created with the reference/key name from Compose), but there are no errors or warnings about that. If we specify it through `volume.external.name`, it works but gives us the deprecation warning. So for now we have to resort to naming our volume through the key.

**Comment by Joyam** [comment]
We use now a 3.4 compose with docker-compose v1.17.1 and docker 17.09-ce and we use `volume.name` without any issue now.
We just needed to remove the `volume.external`.

Hoping to help.

**Comment by dnephin** [comment]
https://github.com/docker/cli/pull/670 also fixd a bug where `volume.name` was not being used for external volumes. The fix will be in Docker v17.12.
